File: eeg-backend/neuro_yogic/connectivity.py
"""
connectivity.py
===============
Non-linear functional connectivity — the reference oracle for the C# port in
``src/NeuroYogic.SignalProcessing/Dsp/Connectivity.cs``.

Weighted Symbolic Mutual Information (wSMI, King et al. 2013)
------------------------------------------------------------
Each channel is transformed into a sequence of ordinal patterns (order k, delay
tau — the same Bandt-Pompe symbolisation used by permutation entropy). For each
channel pair we compute the mutual information between their symbol streams,
*weighting out* symbol pairs that are identical or mirror-images. Those trivial
couplings mostly reflect a common source / volume conduction, so discounting
them leaves genuine non-linear interaction — which distinguished deep absorption
where ordinary phase synchrony (PLV/WPLI) failed.

    wSMI(a,b) = (1/log k!) · Σ w(i,j)·p(i,j)·log[ p(i,j) / (p(i)·p(j)) ]
    w(i,j) = 0 if symbol i == j  or  i == reverse(j),  else 1

Reported as the mean wSMI over all channel pairs (broadband, on the
band-pass+notch signal).
"""
from math import log

import numpy as np


def wsmi(filtered, order: int = 3, delay: int = 1) -> float:
    """Mean weighted symbolic mutual information over all channel pairs."""
    filtered = np.atleast_2d(np.asarray(filtered, dtype=np.float64))
    n_ch = filtered.shape[0]
    syms = [_symbolize(filtered[c], order, delay) for c in range(n_ch)]
    denom = log(_factorial(order))
    vals = []
    for i in range(n_ch):
        for j in range(i + 1, n_ch):
            vals.append(_wsmi_pair(syms[i], syms[j], denom))
    return float(np.mean(vals)) if vals else 0.0


def connectivity_features(filtered) -> dict:
    """Non-linear connectivity summary for one epoch."""
    return {"wsmi": float(wsmi(filtered))}


# ── internals (kept identical to the C# port) ────────────────────────────────

def _symbolize(x, order, delay):
    """Sequence of ordinal-pattern symbols as (code, mirror_code) tuples."""
    n = len(x)
    span = delay * (order - 1)
    out = []
    for i in range(n - span):
        window = [x[i + t * delay] for t in range(order)]
        perm = _argsort_stable(window)
        code = _encode(perm, order)
        mirror = _encode(tuple(reversed(perm)), order)
        out.append((code, mirror))
    return out


def _wsmi_pair(sa, sb, denom):
    t = len(sa)
    if t == 0 or denom <= 0.0:
        return 0.0
    joint: dict = {}
    ca: dict = {}
    cb: dict = {}
    for k in range(t):
        a = sa[k]      # (code, mirror)
        b = sb[k]
        key = (a[0], b[0])
        joint[key] = joint.get(key, 0) + 1
        ca[a[0]] = ca.get(a[0], 0) + 1
        cb[b[0]] = cb.get(b[0], 0) + 1
    # mirror lookup: symbol code -> its mirror code
    mirror_of = {}
    for a in sa:
        mirror_of[a[0]] = a[1]
    for b in sb:
        mirror_of[b[0]] = b[1]

    smi = 0.0
    for (a_code, b_code), cnt in joint.items():
        if a_code == b_code or a_code == mirror_of[b_code]:
            continue
        pab = cnt / t
        pa = ca[a_code] / t
        pb = cb[b_code] / t
        smi += pab * log(pab / (pa * pb))
    return smi / denom


def _argsort_stable(window):
    return tuple(idx for idx, _ in sorted(enumerate(window), key=lambda p: (p[1], p[0])))


def _encode(perm, order):
    code = 0
    mul = 1
    for t in range(order):
        code += perm[t] * mul
        mul *= order
    return code


def _factorial(k):
    f = 1
    for i in range(2, k + 1):
        f *= i
    return f

File: eeg-backend/src/NeuroYogic.SignalProcessing/Dsp/Connectivity.cs
using NeuroYogic.Domain.Analysis;

namespace NeuroYogic.SignalProcessing.Dsp;

/// <summary>
/// Non-linear functional connectivity — faithful C# port of the Python oracle in
/// <c>neuro_yogic/connectivity.py</c>. Computes weighted Symbolic Mutual
/// Information (wSMI, King et al. 2013) over channel pairs: ordinal-pattern
/// symbolisation, mutual information between symbol streams, with identical and
/// mirror-image symbol pairs weighted out to discount volume conduction.
/// </summary>
internal static class Connectivity
{
    public static ConnectivitySet Compute(double[][] filtered) => new(Wsmi(filtered));

    public static double Wsmi(double[][] filtered, int order = 3, int delay = 1)
    {
        var nCh = filtered.Length;
        var mirror = BuildMirrorMap(order);
        var syms = new long[nCh][];
        for (var c = 0; c < nCh; c++) syms[c] = Symbolize(filtered[c], order, delay);

        var denom = Math.Log(Factorial(order));
        double sum = 0.0;
        var pairs = 0;
        for (var i = 0; i < nCh; i++)
            for (var j = i + 1; j < nCh; j++)
            {
                sum += WsmiPair(syms[i], syms[j], denom, mirror);
                pairs++;
            }
        return pairs > 0 ? sum / pairs : 0.0;
    }

    private static double WsmiPair(long[] sa, long[] sb, double denom, IReadOnlyDictionary<long, long> mirror)
    {
        var t = sa.Length;
        if (t == 0 || denom <= 0.0) return 0.0;

        var joint = new Dictionary<(long, long), int>();
        var ca = new Dictionary<long, int>();
        var cb = new Dictionary<long, int>();
        for (var k = 0; k < t; k++)
        {
            var a = sa[k];
            var b = sb[k];
            joint.TryGetValue((a, b), out var jc); joint[(a, b)] = jc + 1;
            ca.TryGetValue(a, out var ac); ca[a] = ac + 1;
            cb.TryGetValue(b, out var bc); cb[b] = bc + 1;
        }

        var smi = 0.0;
        foreach (var kv in joint)
        {
            var (aCode, bCode) = kv.Key;
            if (aCode == bCode || aCode == mirror[bCode]) continue;
            var pab = (double)kv.Value / t;
            var pa = (double)ca[aCode] / t;
            var pb = (double)cb[bCode] / t;
            smi += pab * Math.Log(pab / (pa * pb));
        }
        return smi / denom;
    }

    /// <summary>Ordinal-pattern symbol codes for one channel.</summary>
    private static long[] Symbolize(double[] x, int order, int delay)
    {
        var n = x.Length;
        var span = delay * (order - 1);
        var count = n - span;
        if (count <= 0) return Array.Empty<long>();
        var codes = new long[count];
        var window = new double[order];
        for (var i = 0; i < count; i++)
        {
            for (var t = 0; t < order; t++) window[t] = x[i + t * delay];
            codes[i] = Encode(ArgsortStable(window), order);
        }
        return codes;
    }

    /// <summary>Maps each ordinal-pattern code to the code of its reversed pattern.</summary>
    private static Dictionary<long, long> BuildMirrorMap(int order)
    {
        var map = new Dictionary<long, long>();
        foreach (var perm in Permutations(order))
        {
            var rev = (int[])perm.Clone();
            Array.Reverse(rev);
            map[Encode(perm, order)] = Encode(rev, order);
        }
        return map;
    }

    private static long Encode(int[] perm, int order)
    {
        long code = 0, mul = 1;
        for (var t = 0; t < order; t++) { code += perm[t] * mul; mul *= order; }
        return code;
    }

    private static IEnumerable<int[]> Permutations(int order)
    {
        var items = new int[order];
        for (var i = 0; i < order; i++) items[i] = i;
        return Permute(items, 0);
    }

    private static IEnumerable<int[]> Permute(int[] arr, int start)
    {
        if (start == arr.Length - 1)
        {
            yield return (int[])arr.Clone();
            yield break;
        }
        for (var i = start; i < arr.Length; i++)
        {
            (arr[start], arr[i]) = (arr[i], arr[start]);
            foreach (var p in Permute(arr, start + 1)) yield return p;
            (arr[start], arr[i]) = (arr[i], arr[start]);
        }
    }

    private static int[] ArgsortStable(double[] w)
    {
        var idx = new int[w.Length];
        for (var i = 0; i < w.Length; i++) idx[i] = i;
        Array.Sort(idx, (a, b) =>
        {
            var cmp = w[a].CompareTo(w[b]);
            return cmp != 0 ? cmp : a.CompareTo(b);
        });
        return idx;
    }

    private static long Factorial(int k)
    {
        long f = 1;
        for (var i = 2; i <= k; i++) f *= i;
        return f;
    }
}
